rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,289 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: draft
+---
+
+# RFD 57 - Automatic discovery and enrollment of AWS servers
+
+## What
+
+Proposes a way by which an SSH service might automatically discover and register AWS
+EC2 instances.
+
+## Why
+
+Currently when adding a new AWS server, it's required that Teleport be installed
+after the server has been provisioned which may be a slow process for organizations
+with large numbers of servers as it needs to be installed and then added to the
+teleport cluster
+
+With the changes described in this document, Teleport will be able to resolve the
+issues with adding AWS servers to Teleport clusters automatically.
+
+
+## Discovery
+
+A Teleport SSH agent will need to be configured with an assume role policy for each
+of the accounts that the agent will discover services on as described by the IAM join
+method docs[1]
+
+Discovery can use a matcher similar to the `db_service/aws` matcher, however EC2
+instances will have an optional install command:
+
+```yaml
+ssh_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes" # aws tags to match
+    ssm_command_document: ssm_command_document_name
+    accounts:
+    - aws_account: "222222222222"
+      aws_roles:
+      - "arn:aws:iam::222222222222:role/teleport-DescribeInstances-role"
+      - "arn:aws:iam::222222222222:role/teleport-Install-role"
+```
+
+The agent will use EC2's `DescribeInstances` API in order to list instances[1]. This
+will require the teleport SSH agent to include `ec2:DescribeInstances` as part of
+it's IAM permissions
+
+As with AWS database discover, new EC2 nodes will be discovered periodically on a 60
+second timer, as new nodes are found they will be added to the teleport cluster.
+
+In order to avoid attempting to reinstall teleport on top of an instance where it is
+already present the generated teleport config will include a static label indicating
+it was created via auto discovery.
+
+Example:
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "AWS_INSTANCE_ID",
+    "labels": {
+      "env": "example"
+      "teleport.dev/discovered-node": "yes"
+    },
+  },
+  "spec": {
+    "public_addr": "ec2-54-194-252-215.us-west-1.compute.amazonaws.com",
+    "hostname": "awsxyz"
+  }
+}
+```
+
+IAM join tokens will have a `tags` field added to the accounts so so only EC2
+instances with matching `tags` can be added using the token. the tags matching will
+default to `"*":"*"` to retain compatibility with previously generated IAM tokens
+
+## Agent installation
+
+In order to install the Teleport agent on EC2 instances, the user must create a
+custom SSM Command document. By using the document it allows for only a specific set
+of install commands to be run, rather than having a whole shell available.
+
+Example SSM aws:runCommand document:
+```yaml
+# name: installTeleport
+---
```

Comment:
@r0mant @lxea for teleport discover, can we create this document for them, or can Teleport automatically create a default one so users don't have to think initially about this document but can replace/override it if they want to tweak it?

Reply:
We can do it as a part of `teleport configure bootstrap`? E.g for database access that command sets up IAM policies, for server access it can provision a default SSM document.

Reply:
Also, I'm curious how do we handle versioning. @lxea Can the document command be parametrized? Otherwise how would it determine which Teleport version to install?

Reply:
Another idea would be to serve this document from the cluster itself. We could even do it as a resource which people will be able to edit (kinda like ClusterConfiguration, etc.).

Reply:
@lxea So the difference is this (at least as I understand it, correct me if I'm wrong):

* If we use the ad-hoc script approach, we have to give Teleport IAM permission `arn:aws:ssm:us-west-2:*:instance/*` (from your example below) to allow it to run basically anything on the nodes. If this Teleport instance's IAM creds get compromised, you have permissions to run any script/command on EC2 nodes.

* If we use the `runShellScript` approach, we will give Teleport IAM permission to run _only_ `arn:aws:ssm:us-east-2:aws-account-ID:document/installTeleport` script on the nodes. So if you compromise Teleport's IAM credentials, you can't execute whatever you want, you can only execute `installTeleport` script. And to be able to modify if you need to have additional permissions within Teleport.

Am I missing something? You can assign IAM permissions to run a particular script without also giving permissions to run arbitrary commands, right?